identity-wallet/src/state/actions/load_dev_profile.rs

```diff
@@ -2,12 +2,50 @@ use crate::reducer;
 use crate::state::actions::ActionTrait;
 use crate::state::actions::Reducer;
 use crate::state::reducers::dev_mode::load_dev_profile;
+use ts_rs::TS;
 
-#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
-pub struct LoadDevProfile;
+#[derive(serde::Serialize, serde::Deserialize, Debug, TS, Clone, PartialEq, Eq)]
+#[ts(export, export_to = "bindings/dev/ProfileType.ts")]
+pub enum ProfileType {
+    /// Dev profile with predefined static data
+    Ferris,
 
-#[typetag::serde(name = "[DEV] Load profile")]
-impl ActionTrait for LoadDevProfile {
+    /// Dev profile that can dynamically interact with the NGDIL demo by executing `ProfileSteps`
+    Dragon,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, TS, Clone, PartialEq, Eq, PartialOrd)]
+#[ts(export, export_to = "bindings/dev/ProfileSteps.ts")]
+pub enum ProfileSteps {
+    /// Step 1
+    CreateProfile,
+    /// Step 2
+    AddCredentials,
+    /// Step 3
+    AcceptCredentials,
+    /// Step 4
+    AddConnection,
+    /// Step 5
+    AcceptConnection,
+    /// Step 6
+    AddPresentation,
+    /// Step 7
+    ShareCredentails,
+    /// Step 8
+    AddFutureEngineer,
+    /// Step 9
+    CompleteFlow,
+}
```

Comment:
Since `ProfileSteps` is only relevant for the `Dragon` profile type, should we move this enum to `dynamic_dev_profile.rs`? I'm thinking we should do the same to `pub async fn load_dragon_profile` (move it to `dynamic_dev_profile.rs`). 

Otherwise `ProfileSteps` suggests that its functionality is broader than just the `Dragon` profile type. 

But your design does give me an idea so I'm curious what you think about it (definitely out of scope for this PR though): 

In the future if we want to support more dev profiles (that don't work with NGDIL but against some other parties), what if we make `ProfileSteps` into a trait with an `fn execute` method? All the functions you currently have in `dynamic_dev_profile.rs` e.g:
```rust

pub trait ProfileSteps {
    async fn execute(&self, state: AppState) -> Result<AppState, AppError>;
}

#[derive(serde::Serialize, serde::Deserialize, Debug, TS, Clone, PartialEq, Eq, PartialOrd)]
#[ts(export, export_to = "bindings/dev/ProfileSteps.ts")]
pub enum NGDILProfileSteps {
    /// Step 1
    CreateProfile,
    /// Step 2
    AddCredentials,
    /// Step 3
    AcceptCredentials,
    /// Step 4
    AddConnection,
    /// Step 5
    AcceptConnection,
    /// Step 6
    AddPresentation,
    /// Step 7
    ShareCredentails,
    /// Step 8
    AddFutureEngineer,
    /// Step 9
    CompleteFlow,
}

impl ProfileSteps for NGDILProfileSteps {
    async fn execute(&self, state: AppState) -> Result<AppState, AppError> {
        match self {
            NGDILProfileSteps::CreateProfile => create_new_profile(state).await,
            NGDILProfileSteps::AddCredentials => add_credential(state).await,
            NGDILProfileSteps::AcceptCredentials => accept_credential(state).await,
            NGDILProfileSteps::AddConnection => add_connection(state).await,
            NGDILProfileSteps::AcceptConnection => accept_connection(state).await,
            NGDILProfileSteps::AddPresentation => add_presentation_request(state).await,
            NGDILProfileSteps::ShareCredentails => share_credentials(state).await,
            NGDILProfileSteps::AddFutureEngineer => add_future_engineer(state).await,
            NGDILProfileSteps::CompleteFlow => accept_future_engineer(state).await,
        }
    }
}

impl NGDILProfileSteps {
    pub(super) async fn create_new_profile(&self, state: AppState) -> Result<AppState, AppError> {
        let create_new = CreateNew {
            name: "Shenron".to_string(),
            picture: "&#x1F432".to_string(),
            theme: "dark".to_string(),
            password: PASSWORD.to_string(),
        };

        command::reduce(state, Arc::new(create_new)).await
    }

    pub(super) async fn add_credential(&self, state: AppState) -> Result<AppState, AppError> {
        let state = self.create_new_profile(state).await?; // <-- All steps should execute the 'previous' step first

        // URL from NGDIL demo
        let url = "https://api.demo.ngdil.com/api/starting-offer";

        let payload = serde_json::json!({
            "credentials": [
                "National ID",
                "School Course Certificate",
                "Volunteer Badge",
                "Higher Education Information Literacy Level 1",
                "Business Innovation & Interdisciplinair Samenwerken"
            ]
        });

        let response: CredentialResponse = reqwest::Client::new()
            .post(url)
            .json(&payload)
            .send()
            .await
            .map_err(|err| AppError::DevError(err.to_string()))?
            .json()
            .await
            .map_err(|err| AppError::DevError(err.to_string()))?;

        let qr_code = QrCodeScanned {
            form_urlencoded: response.uri,
        };

        command::reduce(state, Arc::new(qr_code)).await
    }

    ...all the other profile step 'methods'
```

Then this inside `pub async fn load_dev_profile`:
```rust
    if let Some(dev_profile) = listen::<DevProfile>(action) {
        // All dev profiles need to use the const PASSWORD so it can automatically unlock storage.
        match dev_profile.profile {
            ProfileType::Ferris => return load_ferris_profile().await,
            ProfileType::Dragon => return load_dragon_profile(state, dev_profile).await,
        }
    }
```

could just be
```rust
    if let Some(dev_profile) = listen::<DevProfile>(action) {
        return dev_profile.execute().await;
    }
```

Again, just an idea for a future improvement, but interested in what you think :)

Reply:
Ok thanks!

It would work because of this right?:
```rust
    pub(super) async fn add_credential(&self, state: AppState) -> Result<AppState, AppError> {
        let state = self.create_new_profile(state).await?; // <-- All steps should execute the 'previous' step first
```

Reply:
No because of this:

```rust
    if ProfileSteps::AddCredentials <= steps {
        debug!("Add credentials step executed");
        state = add_credential(state).await?;
    }
```
Maybe I should explicitly give the enum a number, but it check if steps (enum 5 is larger then AddCredentials). If so it would also execute that.

Reply:
Using a `match` is just a personal preference but it could also be replaced with your current solution of course:
```rust

    state = reset_settings(state).await?;
    state = create_new_profile(state).await?;

    if self <= steps {
        debug!("Add credentials step executed");
        state = add_credential(state).await?;
    }

    if self <= steps {
        debug!("Accept credentials step executed");
        state = accept_credential(state).await?;
    }

    if self <= steps {
        debug!("Add connection step executed");
        state = add_connection(state).await?;
    }

    if self <= steps {
        debug!("Accept connection step executed");
        state = accept_connection(state).await?;
    }

    if self <= steps {
        debug!("Add presentation step executed");
        state = add_presentation_request(state).await?;
    }

    if self <= steps {
        debug!("Share credentials step executed");
        state = share_credentials(state).await?;
    }

    if self <= steps {
        debug!("Add future engineer step executed");
        state = add_future_engineer(state).await?;
    }

    if self <= steps {
        debug!("Accept future engineer step executed");
        state = accept_future_engineer(state).await?;
    }
```

I was more wondering if you see anything in using a trait in the future for this functionality that could also be implemented for dev profiles outside our own codebase. If so then I will open a discussion for it :)